Undo/COUndoTrackStore.m

```diff
@@ -283,55 +402,180 @@ - (id) deserialize: (NSData *)data
 {
-	[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	assert(dispatch_get_current_queue() != _queue);
+	
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	});
 }
 
 - (BOOL) string: (NSString *)aString matchesGlobPattern: (NSString *)aPattern
 {
-	return [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	assert(dispatch_get_current_queue() != _queue);
+	__block BOOL result = NO;
+
+	dispatch_sync(_queue, ^() {
+		result = [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	});
+	
+	return result;
+}
+
+- (void)markCommandsAsDeletedForUUIDs: (NSArray *)UUIDs
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	
+	__block NSArray *compactedTrackNames = nil;
+
+	// This can be run in background
+	dispatch_sync(_queue, ^() {
+		[_db beginTransaction];
+
+		for (ETUUID *UUID in UUIDs)
+		{
+			[_db executeUpdate: @"UPDATE commands SET deleted = 1 WHERE uuid = ?", [UUID dataValue]];
+		}
+
+		compactedTrackNames =
+			[_db arrayForQuery: @"SELECT DISTINCT trackname FROM commands WHERE deleted = 1 "];
+
+		[_db commit];
+	});
+
+	/* This must be run in the main thread:
+	   - no one must access _modifiedTrackStateForTrackName at the same time
+	   - we must post the commit notification immediately
+	   
+	   If we post it with some delay, someone could touch 
+	   _modifiedTrackStateForTrackName and overwrite COUndoTrackState.compacted. */
+	dispatch_sync_now(dispatch_get_main_queue(), ^() {
+		dispatch_sync(_queue, ^() {
+			[_db beginTransaction];
+
+			for (NSString *trackName in compactedTrackNames)
+			{
+				if (_modifiedTrackStateForTrackName[trackName] == nil)
+				{
+					_modifiedTrackStateForTrackName[trackName] = [self stateForTrackNameInCurrentQueue: trackName];
+				}
+				((COUndoTrackState *)_modifiedTrackStateForTrackName[trackName]).compacted = YES;
+			}
+
+			[_db commit];
+		});
+
+		[self postCommitNotifications];
+	});
+
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+- (void)finalizeDeletions
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE deleted = 1"];
+	});
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+/**
+ * We run vacuum with dispatch_sync() in a queue, so we can sure no other
+ * threads access the database when we invoke -vacuum in a background thread.
+ */
+- (BOOL)vacuum
+{
+    assert(dispatch_get_current_queue() != _queue);
+	__block BOOL success = NO;
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
```

Comment:
I tried to implement -markCommandsAsDeletedForUUIDs:withCompletionHandler:, -finalizeDeletionsWithCompletionHandler and -vacuumWithCompletionHandler, but it doesn't work out, because this requires to change all calling APIs to use a completion handler.

So -[COHistoryCompaction beginCompaction] needs to be turned into -beingCompactionWithCompletionHandler:. In the same way, -[COSQLiteStore compactHistory:] needs to be rewritten into -compactHistoryWithCompletionHandler:, but then this means changing more COSQLiteStore methods and introducing some asynchronous API (e.g. -finalizeDeletionsForPersistentRoot:error:withCompletionHandler:). This doesn't sound like a good idea, since COSQLiteStore API is designed as an entirely synchronous API.

I now remember that I also wrapped COUndoTrackStore background operations such as vacuum with a synchronous API in order to mimick COSQLiteStore API.

So I'm not sure what to do… I'm tempted to say that we should keep both COSQLiteStore and COUndoTrackStore APIs purely synchronous (to keep the API symetric). Using completion handlers aslo translate into less readable implementation code due to the nested callbacks. 

May be we should postpone this change?